cel/library.go

```diff
@@ -439,119 +435,48 @@ func (l library) validateEmail(addr string) bool {
 	return l.emailRegex.MatchString(addr)
 }
 
+// validateHostname validates whether host is a valid hostname.
 func (l library) validateHostname(host string) bool {
-	if len(host) > 253 {
-		return false
-	}
-
-	s := strings.ToLower(strings.TrimSuffix(host, "."))
-	allDigits := false
-	// split hostname on '.' and validate each part
-	for _, part := range strings.Split(s, ".") {
-		allDigits = true
-		// if part is empty, longer than 63 chars, or starts/ends with '-', it is invalid
-		if l := len(part); l == 0 || l > 63 || part[0] == '-' || part[l-1] == '-' {
-			return false
-		}
-		// for each character in part
-		for _, ch := range part {
-			// if the character is not a-z, 0-9, or '-', it is invalid
-			if (ch < 'a' || ch > 'z') && (ch < '0' || ch > '9') && ch != '-' {
-				return false
-			}
-			allDigits = allDigits && ch >= '0' && ch <= '9'
-		}
-	}
-
-	// the last part cannot be all numbers
-	return !allDigits
+	return isHostname(host)
 }
 
+// validateIP validates whether addr is a valid IP address for the given version.
+// If ver is 4, it will validate str as an ipv4 address. If ver is 6,
+// it will validate as ipv6. If ver is 0, it will validate that str is
+// _either_ ipv4 or ipv6.
+//
+// If ver is any value other than 4, 6, or 0, returns false.
 func (l library) validateIP(addr string, ver int64) bool {
-	address := net.ParseIP(addr)
-	if address == nil {
-		return false
-	}
-	switch ver {
-	case 0:
-		return true
-	case 4:
-		return address.To4() != nil
-	case 6:
-		return address.To4() == nil
-	default:
-		return false
-	}
+	return isIP(addr, ver)
```

Comment:
Instead of calling `l.validateIP(addr, 0)` in the library, why don't we call `isIp(addr, 0)` instead, and remove this method? It only adds indirection.

Same for all other similar cases. 


Somewhat related: uri.go defines functions that aren't used for URIs, for example `isIPPrefix` and `isHostAndPort`. I think we can simply move all definitions into library.go, especially if we remove the library methods.